lib/plugins/aws/invokeLocal/index.js

```diff
@@ -300,30 +316,62 @@ class AwsInvokeLocal {
     );
   }
 
+  getDockerImageName() {
+    return `sls-docker-${this.getRuntime()}`;
+  }
+
   buildDockerImage(layerPaths) {
     const runtime = this.getRuntime();
-    const imageName = 'sls-docker';
+    const imageName = this.getDockerImageName();
 
-    return new BbPromise((resolve, reject) => {
-      let dockerfile = `FROM lambci/lambda:${runtime}`;
-      for (const layerPath of layerPaths) {
-        dockerfile += `\nADD --chown=sbx_user1051:495 ${layerPath} /opt`;
-      }
-      mkdirp.sync(path.join('.serverless', 'invokeLocal'));
-      const dockerfilePath = path.join('.serverless', 'invokeLocal', 'Dockerfile');
-      fs.writeFileSync(dockerfilePath, dockerfile);
-      this.serverless.cli.log('Building Docker image...');
-      const docker = spawn('docker', [
-        'build',
-        '-t',
-        imageName,
-        `${this.serverless.config.servicePath}`,
-        '-f',
-        dockerfilePath,
-      ]);
-      docker.on('exit', error => {
-        return error ? reject(error) : resolve(imageName);
+    let dockerfile = `FROM lambci/lambda:${runtime}`;
+    for (const layerPath of layerPaths) {
+      dockerfile += `\nADD --chown=sbx_user1051:495 ${layerPath} /opt`;
+    }
+
+    const dockerfilePath = path.join('.serverless', 'invokeLocal', runtime, 'Dockerfile');
+    const dockerfileCachePath = path.join(cachePath, 'dockerfiles', runtime, 'Dockerfile');
+
+    let cachedDockerfileEquals = BbPromise.resolve();
+
+    if (this.serverless.utils.fileExistsSync(dockerfileCachePath)) {
+      cachedDockerfileEquals = fs.readFileAsync(dockerfileCachePath).then(contents => {
+        return contents.toString() === dockerfile && this.checkDockerImage(imageName);
       });
+    }
+
+    return cachedDockerfileEquals.then(cacheEquals => {
+      if (cacheEquals) {
+        return imageName;
+      }
+      return BbPromise.all([
+        mkdirp.async(path.join('.serverless', 'invokeLocal', runtime)),
+        mkdirp.async(path.join(cachePath, 'dockerfiles', runtime)),
+      ])
+        .then(() => {
+          this.serverless.cli.log('Writing Dockerfile...');
+          return BbPromise.all([
+            fs.writeFileSync(dockerfilePath, dockerfile),
+            fs.writeFileSync(dockerfileCachePath, dockerfile),
+          ]);
+        })
+        .then(
+          () =>
+            new BbPromise((resolve, reject) => {
+              this.serverless.cli.log('Building Docker image...');
+              const docker = spawn('docker', [
```

Comment:
This is a dev dependency only and used in tests/scripts only so will keep original node spawn :)

Reply:
It is a normal dependency, ensure to be up to date with master :)

Reply:
In this case, however, where we don't rely on the output and listen for exit in most cases it kinds of defeats the purpose :) Please check the updated PR :)

Reply:
> it kinds of defeats the purpose 

Not really, it provides promise support and eventual error debugging out of a box. So there's need for creating promise wrapper if you use that.

We definitely should rely on it.

Reply:
@medikoo I started implementing this but it caused a lot of issues, and a lot of stuff broke in tests etc where child process is stubbed. I gave it a try but I can't get it right. Please check out my branch https://github.com/richarddd/serverless/tree/use-child-process-ext which is based.

Maybe we could merge this and then do another pr with your suggested optimization or you could take a look at my code and figure out what I'm doing wrong? ☺️

Reply:
Indeed as it's function that's directly exported you can't stub with sinon easily. But you can do it via `proxyquire`, as e.g. it's done here: https://github.com/serverless/serverless/blob/db4ff0b1cf62f371c610b31666c7872698b2bc77/lib/utils/downloadTemplateFromRepo.test.js#L61

Reply:
I gave that a try as well with no success :( The spawn function returned by use-child-process-ext is a promised wrapped and we cant simply stub its in: https://github.com/serverless/serverless/blob/db4ff0b1cf62f371c610b31666c7872698b2bc77/lib/utils/downloadTemplateFromRepo.test.js#L61

I just have a feeling this PR is getting out of scope:

https://github.com/serverless/serverless/blob/a3f691e369777d07ae6dff104e7d8ef50e7391dd/lib/plugins/aws/invokeLocal/index.test.js#L986

Reply:
@medikoo could you give me a hit of what im doing wrong here? I'm not very experianced with sinon:

```
writeChildStub = sinon.stub();
endChildStub = sinon.stub();
stdinStub = sinon.stub().resolves('');
spawnStub = sinon.stub().returns({
  stderr: new EventEmitter().on('data', () => {}),
  stdout: new EventEmitter().on('data', () => {}),
  stdin: {
    write: writeChildStub,
    end: endChildStub,
  },
  on: (key, callback) => {
    if (key === 'close') process.nextTick(callback);
  },
  then: () => {},
});
AwsInvokeLocal = proxyquire('./index', {
  'get-stdin': stdinStub,
  'child-process-ext/spawn': spawnStub,
});
```

Reply:
If you need to read output of the spawned process. Then _promise_ resolves with it. See the doc: https://github.com/medikoo/child-process-ext/tree/60e5781d2c848186c9727e8c638afdee73bcb3a6#spawncommand-args-options

Stub that provides command output can be configured as:

```javascript
AwsInvokeLocal = proxyquire('./index', {
  'get-stdin': stdinStub,
  'child-process-ext/spawn': sinon.stub().resolves({
    stdoutBuffer: new Buffer("Mocked output")
  });
});

```

Reply:
> @medikoo not only that since child-process-ext/spawn returns a "decorated promise".

That's true, but I believe there's no need to rely on that _decoration_ in a plugin, so you can ignore it.

Reply:
> > @medikoo not only that since child-process-ext/spawn returns a "decorated promise".
> 
> That's true, but I believe there's no need to rely on that _decoration_ in a plugin, so you can ignore it.

Actually it is. We want the invoked lambdas output to write to `process.stdout` in realtime. We don't want to wait for the process to exit before printing the result :)

Reply:
> We want the invoked lambdas output to write to process.stdout in realtime

Where exactly? In scope of this PR, I see two usages of `spawn`:

1. https://github.com/serverless/serverless/blob/f76f65ef69cb5dfa9b77c2eebc708b068d47d933/lib/plugins/aws/invokeLocal/index.js#L241
2. https://github.com/serverless/serverless/blob/f76f65ef69cb5dfa9b77c2eebc708b068d47d933/lib/plugins/aws/invokeLocal/index.js#L377

And none of them seem to be configured to provide live output

>  Can't be done because of invocations with python, java 

This I think is not related to docker handling (so this PR). I would leave that code not touched